File: api/app/api/schemas/research.py
"""Deep-research request and response contracts."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ResearchCreate(BaseModel):
    question: str = Field(min_length=5, max_length=8000)
    title: str | None = Field(default=None, min_length=1, max_length=256)
    knowledge_base_ids: list[uuid.UUID] = Field(default_factory=list, max_length=20)
    use_memory: bool = True
    allow_web: bool = False
    max_iterations: int = Field(default=2, ge=1, le=3)


class ResearchStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    position: int
    question: str
    status: str
    finding: str | None
    evidence_count: int
    started_at: datetime | None
    finished_at: datetime | None
    error_message: str | None


class ResearchEvidenceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    step_id: uuid.UUID | None
    source_type: str
    source_id: str
    chunk_id: str | None
    title: str
    quote: str
    url: str | None
    locator: dict[str, object] | None
    score: float | None
    query: str


class ResearchTaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    question: str
    status: str
    stage: str
    progress: float
    allow_web: bool
    use_memory: bool
    knowledge_base_ids: list[str]
    plan: dict[str, object] | None
    verifier_result: dict[str, object] | None
    report_markdown: str | None
    iteration_count: int
    max_iterations: int
    input_tokens: int
    output_tokens: int
    total_tokens: int
    started_at: datetime | None
    finished_at: datetime | None
    error_code: str | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime


class ResearchTaskDetail(ResearchTaskResponse):
    steps: list[ResearchStepResponse] = Field(default_factory=list)
    evidence: list[ResearchEvidenceResponse] = Field(default_factory=list)

File: web/src/entities/research.ts
export type ResearchStatus =
  | "pending"
  | "retrying"
  | "planning"
  | "researching"
  | "verifying"
  | "writing"
  | "completed"
  | "failed";

export type ResearchTask = {
  id: string;
  title: string;
  question: string;
  status: ResearchStatus;
  stage: string;
  progress: number;
  allow_web: boolean;
  use_memory: boolean;
  knowledge_base_ids: string[];
  plan: { objective?: string; questions?: string[]; deliverables?: string[] } | null;
  verifier_result: {
    sufficient?: boolean;
    coverage?: number;
    gaps?: string[];
    conflicts?: string[];
    follow_up_queries?: string[];
  } | null;
  report_markdown: string | null;
  iteration_count: number;
  max_iterations: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  started_at: string | null;
  finished_at: string | null;
  error_code: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
};

export type ResearchStep = {
  id: string;
  position: number;
  question: string;
  status: string;
  finding: string | null;
  evidence_count: number;
  started_at: string | null;
  finished_at: string | null;
  error_message: string | null;
};

export type ResearchEvidence = {
  id: string;
  step_id: string | null;
  source_type: string;
  source_id: string;
  chunk_id: string | null;
  title: string;
  quote: string;
  url: string | null;
  locator: Record<string, unknown> | null;
  score: number | null;
  query: string;
};

export type ResearchTaskDetail = ResearchTask & {
  steps: ResearchStep[];
  evidence: ResearchEvidence[];
};

export type ResearchCreate = {
  question: string;
  title?: string;
  knowledge_base_ids: string[];
  use_memory: boolean;
  allow_web: boolean;
  max_iterations: number;
};
